python/ray/util/client/dataclient.py

```diff
@@ -59,48 +59,122 @@ def _start_datathread(self) -> threading.Thread:
+                stub = ray_client_pb2_grpc.RayletDataStreamerStub(
+                    self.client_worker.channel)
+                metadata = self._metadata + \
+                    [("reconnecting", str(reconnecting))]
+                resp_stream = stub.Datapath(
+                    iter(self.request_queue.get, None),
+                    metadata=metadata,
+                    wait_for_ready=True)
+                try:
+                    for response in resp_stream:
+                        self._process_response(response)
+                    return
+                except grpc.RpcError as e:
+                    reconnecting = self._process_rpc_error(e)
+                    if not reconnecting:
+                        break
+                    else:
+                        self._reconnect_channel()
+        finally:
+            logger.info("Shutting down data channel")
+            self._shutdown()
+
+    def _process_response(self, response: Any) -> None:
+        """
+        Process responses from the data servicer.
+        """
+        if response.req_id == 0:
+            # This is not being waited for.
+            logger.debug(f"Got unawaited response {response}")
+            return
+        if response.req_id in self.asyncio_waiting_data:
+            # Async response. Check for callback and update outstanding
+            # requests.
+            callback = self.asyncio_waiting_data.pop(response.req_id)
+            try:
+                if callback:
+                    callback(response)
+            except Exception:
+                logger.exception("Callback error:")
+            with self.lock:
+                if response.req_id in self.outstanding_requests:
+                    del self.outstanding_requests[response.req_id]
+                    self._acknowledge(response.req_id)
+        else:
+            # Blocking response. Populate ready data and wake up waiting
+            # threads.
             with self.cv:
-                self._in_shutdown = True
+                self.ready_data[response.req_id] = response
                 self.cv.notify_all()
-            if e.code() == grpc.StatusCode.CANCELLED:
-                # Gracefully shutting down
-                logger.info("Cancelling data channel")
-            elif e.code() in (grpc.StatusCode.UNAVAILABLE,
-                              grpc.StatusCode.RESOURCE_EXHAUSTED):
-                # TODO(barakmich): The server may have
-                # dropped. In theory, we can retry, as per
-                # https://grpc.github.io/grpc/core/md_doc_statuscodes.html but
-                # in practice we may need to think about the correct semantics
-                # here.
-                logger.info("Server disconnected from data channel")
-            else:
-                logger.exception(
-                    "Got Error from data channel -- shutting down:")
+
+    def _acknowledge(self, req_id: int):
+        self.request_queue.put(
```

Comment:
I played around with the piggy back approach and ran into a sketchy situation where a bunch of release requests were backed up in the cache, and no more requests were being sent. This isn't a big deal right now, but if for some reason we're caching large async responses we could end up with a large cache that doesn't get cleared out, so I'd rather have some kind of bound on the size of the cache. Will compromise by only sending the ack on every 10th response, but that can be adjusted.

Also, I'm not actually sure how much extra overhead it would be, since the stream is constantly open anyway. If every 10th is still too much overhead, we can always adjust down the line.